pkg/webhook/admission/pod/agent_injector.go

```diff
@@ -0,0 +1,188 @@
+	if ok {
+		args = append(args, constants.AgentS3EndpointArgName)
+		args = append(args, s3Endpoint)
+	}
+
+	// Make sure securityContext is initialized and valid
+	securityContext := pod.Spec.Containers[0].SecurityContext.DeepCopy()
+
+	agentContainer := &v1.Container{
+		Name:  constants.AgentContainerName,
+		Image: il.config.Image,
+		Args:  args,
+		Resources: v1.ResourceRequirements{
+			Limits: map[v1.ResourceName]resource.Quantity{
+				v1.ResourceCPU:    resource.MustParse(il.config.CpuLimit),
+				v1.ResourceMemory: resource.MustParse(il.config.MemoryLimit),
+			},
+			Requests: map[v1.ResourceName]resource.Quantity{
+				v1.ResourceCPU:    resource.MustParse(il.config.CpuRequest),
+				v1.ResourceMemory: resource.MustParse(il.config.MemoryRequest),
+			},
+		},
+		SecurityContext: securityContext,
+	}
+
+	// Add container to the spec
+	pod.Spec.Containers = append(pod.Spec.Containers, *agentContainer)
+
+	if _, ok := pod.ObjectMeta.Annotations[constants.AgentInternalAnnotationKey]; ok {
+		// Mount the modelDir volume to the pod and model agent container
+		err := mountModelDir(pod)
+		if err != nil {
+			return err
+		}
+		// Mount the modelConfig volume to the pod and model agent container
+		err = mountModelConfig(pod)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func mountModelDir(pod *v1.Pod) error {
+	if _, ok := pod.ObjectMeta.Annotations[constants.AgentModelDirAnnotationKey]; ok {
+		modelDirVolume := v1.Volume{
+			Name: constants.ModelDirVolumeName,
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		}
+		mountVolume(constants.AgentContainerName, pod, modelDirVolume, constants.ModelDir)
+		return nil
+	}
+	return fmt.Errorf("can not find %v label", constants.AgentModelConfigAnnotationKey)
+}
+
+func mountModelConfig(pod *v1.Pod) error {
+	if modelConfigName, ok := pod.ObjectMeta.Annotations[constants.AgentModelConfigAnnotationKey]; ok {
+		modelConfigVolume := v1.Volume{
+			Name: constants.ModelConfigVolumeName,
+			VolumeSource: v1.VolumeSource{
+				ConfigMap: &v1.ConfigMapVolumeSource{
+					LocalObjectReference: v1.LocalObjectReference{
+						Name: modelConfigName,
+					},
+				},
+			},
+		}
+		mountVolume(constants.AgentContainerName, pod, modelConfigVolume, constants.ModelConfigDir)
+		return nil
+	}
+	return fmt.Errorf("can not find %v label", constants.AgentModelConfigAnnotationKey)
+}
+
+func mountVolume(containerName string, pod *v1.Pod, additionalVolume v1.Volume, mountPath string) {
+	pod.Spec.Volumes = appendVolume(pod.Spec.Volumes, additionalVolume)
+	var mountedContainers []v1.Container
+	for _, container := range pod.Spec.Containers {
+		if container.Name == containerName {
+			if container.VolumeMounts == nil {
+				container.VolumeMounts = []v1.VolumeMount{}
+			}
+			container.VolumeMounts = append(container.VolumeMounts, v1.VolumeMount{
+				Name:      additionalVolume.Name,
+				ReadOnly:  true,
```

Comment:
I think model directory is not read only?

Reply:
Agree, should set it to false.